python/cuspatial/cuspatial/core/binpreds/binpreds.py

```diff
@@ -168,23 +185,71 @@ def _op(self, lhs, points):
-            lhs.polygons.ring_offset,
-            lhs.polygons.x,
-            lhs.polygons.y,
+    def _postprocess_quadtree_result(self, point_indices, point_result):
+        # If there are more than 31 polygons, the point indices are
+        # returned as a dataframe with two columns: part_index and
+        # point_index.
+
+        # This complex block of code is to create a dataframe that
+        # contains the polygon index and the point index for each
+        # point in the polygon. Quadtree pip returns the _part_index_
+        # of the polygon, not the polygon index.
+        rings_to_parts = cp.array(self.lhs.polygons.part_offset)
+        part_sizes = rings_to_parts[1:] - rings_to_parts[:-1]
+        parts_map = cudf.Series(
+            cp.arange(len(part_sizes)), name="part_index"
+        ).repeat(part_sizes)
+        parts_df = parts_map.reset_index(drop=True).reset_index()
+        # Mapping of parts to polygons
+        parts_to_geoms = cp.array(self.lhs.polygons.geometry_offset)
+        geometry_sizes = parts_to_geoms[1:] - parts_to_geoms[:-1]
+        geometry_map = cudf.Series(
+            cp.arange(len(geometry_sizes)), name="polygon_index"
+        ).repeat(geometry_sizes)
+        geom_df = geometry_map.reset_index(drop=True)
+        geom_df.index.name = "part_index"
+        geom_df = geom_df.reset_index()
+
+        # Replace the part index with the polygon index
+        part_result = parts_df.merge(point_result, on="part_index")
+        # Replace the polygon index with the row index
+        result = geom_df.merge(part_result, on="part_index")
+        result = result[["polygon_index", "point_index"]]
+        result = result.drop_duplicates()
+        # Replace the polygon index with the original index
+        result["polygon_index"] = result["polygon_index"].replace(
+            cudf.Series(self.lhs.index, index=cp.arange(len(self.lhs.index)))
         )
-        return point_result
+        # Using allpairs for all requests with more than 31 polygons.
+        if not self.allpairs:
+            if len(result) == 0:
+                return cudf.Series([False] * len(self.lhs))
+            final_result = cudf.Series([False] * len(point_indices))
+            grouped = result.groupby("polygon_index").count() == len(
+                point_indices
+            )
+            final_result.loc[grouped.index] = True
+            return final_result
+        else:
+            return result
 
-    def postprocess(self, point_indices, point_result):
-        """Postprocess the output GeoSeries to ensure that they are of the
-        correct type for the predicate."""
-        result = cudf.DataFrame({"idx": point_indices, "pip": point_result})
-        df_result = result
+    def _postprocess_brute_force_result(self, point_indices, point_result):
+        # If there are 31 or fewer polygons in the input, the result
+        # is a dataframe with one row per point and one column per
+        # polygon.
```

Comment:
One option would just be to change this. For the small-polygon-count version, just convert the bitmask into the same format output as the quadtree (all pairs): one row per intersection, containing a polygon index and a point index.

Then contains_properly only needs to convert the results if the user requests pairwise instead of all-pairs.

Reply:
This is a valid option but I can't think of a way to do it with `cudf` operations and no iteration.

Reply:
Turns out it is really easy to change the output format with a `cudf.stack()` call, so I'm investigating this approach to reduce code-paths.

Reply:
I've finished that refactor, it makes solving the discrete math problem at the end harder but I'm ok with it for the current iteration of the architecture.